File: Day4/part1/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

// Read input from file and convert it to a 2D rune slice
func readFile() [][]rune {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var input [][]rune
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		text := scanner.Text()
		input = append(input, []rune(text))
	}

	return input
}

var directions = [8][2]int{
	{-1, 0},  // up
	{1, 0},   // down
	{0, -1},  // left
	{0, 1},   // right
	{-1, -1}, // up-left
	{-1, 1},  // up-right
	{1, -1},  // down-left
	{1, 1},   // down-right
}

// Check if coordinates are within bounds
func checkBounds(rows, cols, x, y int) bool {
	return x >= 0 && x < rows && y >= 0 && y < cols
}

// Search for a specific word in all directions
func searchWord(grid [][]rune, word string) int {
	rows := len(grid)
	cols := len(grid[0])
	matches := 0

	for x, row := range grid {
		for y := range row {
			for _, d := range directions {
				match := true
				for i := 0; i < len(word); i++ {
					newX := x + i*d[0]
					newY := y + i*d[1]
					if !checkBounds(rows, cols, newX, newY) || grid[newX][newY] != rune(word[i]) {
						match = false
						break
					}
				}
				if match {
					matches++
				}
			}
		}
	}

	return matches
}

// Part 1: Search for occurrences of "XMAS"
func part1() {
	grid := readFile()
	matches := searchWord(grid, "XMAS")
	fmt.Println("Part 1 answer:", matches)
}

func main() {
	part1()
}

File: Day4/part1/main.py
def read_file():
    """Reads the input file and returns a 2D list of characters."""
    with open("input.txt", "r") as file:
        return [list(line.strip()) for line in file]

DIRECTIONS = [
    (-1, 0),  # up
    (1, 0),   # down
    (0, -1),  # left
    (0, 1),   # right
    (-1, -1), # up-left
    (-1, 1),  # up-right
    (1, -1),  # down-left
    (1, 1),   # down-right
]

def check_bounds(rows, cols, x, y):
    """Checks if the coordinates (x, y) are within the grid boundaries."""
    return 0 <= x < rows and 0 <= y < cols

def search_word(grid, word):
    """Searches for occurrences of the word in the grid in all 8 directions."""
    rows = len(grid)
    cols = len(grid[0])
    matches = 0

    for x in range(rows):
        for y in range(cols):
            for dx, dy in DIRECTIONS:
                match = True
                for i in range(len(word)):
                    new_x = x + i * dx
                    new_y = y + i * dy
                    if not check_bounds(rows, cols, new_x, new_y) or grid[new_x][new_y] != word[i]:
                        match = False
                        break
                if match:
                    matches += 1
    return matches

def part1():
    grid = read_file()
    matches = search_word(grid, "XMAS")
    print("Part 1 answer:", matches)

if __name__ == "__main__":
    part1()
